pkg/deployment/reconcile/plan_executor.go

```diff
@@ -116,40 +116,76 @@ func (p plannerResources) Set(deployment *api.DeploymentStatus, plan api.Plan) b
 // Returns true when it has to be called again soon.
 // False otherwise.
 func (d *Reconciler) ExecutePlan(ctx context.Context) (bool, error) {
+	execution := 0
+
+	var retrySoon bool
+
+	for {
+		if execution >= 32 {
+			return retrySoon, nil
+		}
+
+		execution++
+
+		if retrySoonCall, recall, err := d.executePlanInLoop(ctx); err != nil {
+			return false, err
+		} else {
+			retrySoon = retrySoon || retrySoonCall
+			if recall {
+				continue
+			}
+		}
+
+		break
+	}
+
+	return retrySoon, nil
+}
+
+// ExecutePlan tries to execute the plan as far as possible.
+// Returns true when it has to be called again soon.
+// False otherwise.
+func (d *Reconciler) executePlanInLoop(ctx context.Context) (bool, bool, error) {
 	var callAgain bool
+	var callInLoop bool
 
-	if again, err := d.executePlanStatus(ctx, plannerHigh{}); err != nil {
-		return false, errors.WithStack(err)
-	} else if again {
-		callAgain = true
+	if again, inLoop, err := d.executePlanStatus(ctx, plannerHigh{}); err != nil {
+		d.planLogger.Err(err).Error("Execution of plan failed")
+		return false, false, errors.WithStack(err)
+	} else {
+		callAgain = callAgain || again
+		callInLoop = callInLoop || inLoop
 	}
 
-	if again, err := d.executePlanStatus(ctx, plannerResources{}); err != nil {
+	if again, inLoop, err := d.executePlanStatus(ctx, plannerResources{}); err != nil {
 		d.planLogger.Err(err).Error("Execution of plan failed")
-		return false, nil
-	} else if again {
-		callAgain = true
+		return false, false, nil
+	} else {
+		callAgain = callAgain || again
+		callInLoop = callInLoop || inLoop
 	}
 
-	if again, err := d.executePlanStatus(ctx, plannerNormal{}); err != nil {
-		return false, errors.WithStack(err)
-	} else if again {
-		callAgain = true
+	if again, inLoop, err := d.executePlanStatus(ctx, plannerNormal{}); err != nil {
+		d.planLogger.Err(err).Error("Execution of plan failed")
+		return false, false, errors.WithStack(err)
+	} else {
+		callAgain = callAgain || again
+		callInLoop = callInLoop || inLoop
 	}
 
-	return callAgain, nil
+	return callAgain, callInLoop, nil
 }
 
-func (d *Reconciler) executePlanStatus(ctx context.Context, pg planner) (bool, error) {
+func (d *Reconciler) executePlanStatus(ctx context.Context, pg planner) (bool, bool, error) {
 	loopStatus, _ := d.context.GetStatus()
 
 	plan := pg.Get(&loopStatus)
 
 	if len(plan) == 0 {
-		return false, nil
+		return false, false, nil
 	}
 
-	newPlan, callAgain, err := d.executePlan(ctx, plan, pg)
+	newPlan, callAgain, callInLoop, err := d.executePlan(ctx, plan, pg)
```

Comment:
What is the intension of `callInLoop` if it is always `false`?

Reply:
Please, check latest code - it can be true